framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,531 @@
+        if (enumConstants.containsKey(classname) && "".equals(rest)) {
+            printWriter.print("enum");
+        } else {
+            printWriter.print("class ");
+        }
+        printAnnotations(aClass.tlAnnotationsHere);
+        printWriter.print(" " + nameToPrint);
+        printTypeParameters(classname);
+        printWriter.println(" {");
+        if ("".equals(rest)) {
+            return 0;
+        } else {
+            return 1 + printClassDefinitions(rest, classname, aClass);
+        }
+    }
+
+    /**
+     * The implementation of {@link #write(AScene, Map, Map, Map, Writer)} and {@link #write(AScene,
+     * Map, Map, Map, String)}. Prints imports, classes, method signatures, and fields in stub file
+     * format, all with appropriate annotations.
+     *
+     * @param scene the scene to write
+     */
+    private void writeImpl(AScene scene) {
+        // Write out all imports
+        ImportDefCollector importDefCollector;
+        try {
+            importDefCollector = new ImportDefCollector(scene);
+        } catch (DefException e) {
+            throw new BugInCF(e.getMessage(), e);
+        }
+        importDefCollector.visit();
+        printWriter.println();
+
+        // For each class
+        for (Map.Entry<String, AClass> classEntry : scene.classes.entrySet()) {
+            String classname = classEntry.getKey();
+            AClass aClass = classEntry.getValue();
+            String pkg = packagePart(classname);
+            if (!"".equals(pkg)) {
+                printWriter.println("package " + pkg + ";");
+            }
+            String basename = basenamePart(classname);
+            if ("package-info".equals(basename) || "module-info".equals(basename)) {
+                continue;
+            }
+
+            // At this point, we no longer care about the distinction between packages
+            // and inner classes, so we should replace the $ in the definition of any
+            // inner classes with a ., so that they are printed correctly in stub files.
+            classname = classname.replace('$', '.');
+
+            int curlyCount = 1 + printClassDefinitions(basename, classname, aClass);
+
+            // print fields or enum constants
+            if (!enumConstants.containsKey(classname)) {
+                for (Map.Entry<String, AField> fieldEntry : aClass.fields.entrySet()) {
+                    String fieldName = fieldEntry.getKey();
+                    AField aField = fieldEntry.getValue();
+                    printWriter.println();
+                    printWriter.print(INDENT);
+                    printAField(aField, fieldName, classname);
+                    printWriter.println(";");
+                }
+            } else {
+                // for enums, instead of printing fields print the enum constants
+                List<VariableElement> enumConstants = enumConstants.get(classname);
+                boolean first = true;
```

Comment:
Here and elsewhere, use StringJoiner or similar to avoid having to write the `first` logic.  That will make the code shorter, simpler, and easier to understand.